File: src_cpp/lexing/character_parser.cpp
#include <string>
#include "character_parser.h"

bool CharacterParser::is_number(char &c)
{
    return c >= 48 && c <= 57;
}

bool CharacterParser::is_letter(char &c)
{
    return c >= 65 && c <= 90 || c >= 97 && c <= 122;
}

bool CharacterParser::is_underscore(char &c)
{
    return c == 95;
}

bool CharacterParser::is_whitespace(char &c)
{
    return isspace(c);
}

bool CharacterParser::is_symbol(char &c)
{
    return c >= 33 && c <= 47 || c >= 58 && c <= 64 || c >= 92 && c <= 96 || c >= 123 && c <= 126;
}

File: src/token/character_parser.ts
function toChar(char: string): number {
    if (SharedArrayBuffer.length != 1) {
        throw new Error("char must be 1 character");
    }

    return char.charCodeAt(0);
}

export class CharacterParser {
    static is_number(c: string) {
        return toChar(c) >= 48 && toChar(c) <= 57;
    }

    static is_letter(c: string) {
        return (toChar(c) >= 65 && toChar(c) <= 90) || (toChar(c) >= 97 && toChar(c) <= 122);
    }

    static is_underscore(c: string) {
        return toChar(c) == 95;
    }

    static is_whitespace(c: string) {
        return c == " " || c == "\n" || c == "\t" || c == "\r";
    }

    static is_symbol(c: string) {
        return (
            (toChar(c) >= 33 && toChar(c) <= 47) ||
            (toChar(c) >= 58 && toChar(c) <= 64) ||
            (toChar(c) >= 92 && toChar(c) <= 96) ||
            (toChar(c) >= 123 && toChar(c) <= 126)
        );
    }
}
